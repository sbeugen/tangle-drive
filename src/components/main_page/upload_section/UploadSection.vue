<template>
  <div>
    <p>1. Select a file you want to upload to the Tangle.</p>
    <div class="file-input-grp">
      <input class="file-input" type="input" v-model="getFileFromState.name" placeholder="Select a File" disabled="true">
      <input type="file" id="file-selector" @click="fileInputClickHandler()" @change="setSelectedFile()" hidden> <!--This input is needed to open the file picker dialog-->
      <button @click="openFilePicker()" :disabled="disabled">File</button>
      <br>
      <br>
      <template v-if="getFileFromState">
        <p>2. Upload your file.</p>
        <button @click="uploadClickHandler()" :disabled="uploadDisabled">Upload</button>
      </template>
      <template v-if="getBundleHashFromState">
        <p>3. This is your bundle hash. Share it with people you want to download your file.</p>
        <input class="bundle-hash" id="bundleHash" type="text" v-model="getBundleHashFromState" placeholder="Bundle Hash" @click="bundleClickHandler()">
      </template>
    </div>
  </div>
</template>

<script>
  import { mapActions, mapGetters } from 'vuex'

  export default {
    data() {
      return {
        disabled: false,
        uploadDisabled: false,
        MAX_FILE_SIZE: 105000000
      }
    },
    methods: { 
      openFilePicker() {
        document.getElementById('file-selector').click()
      },
      fileInputClickHandler() {
        return null
      },
      setSelectedFile(event) {
        if (event.target.files.length > 0) {
          if (!(event.target.files[0].size > this.MAX_FILE_SIZE)){
            this.setFileToState(event.target.files[0])
            this.resetBundleHash()
            this.uploadDisabled = false
          } else {
            alert('The selected file is to big. Currently it is only allowed to upload 100 MB per file.')
          }
        }
      },
      async uploadClickHandler() {
        try {
          this.disabled = true
          this.uploadDisabled = true
          await this.uploadFileToTangle(this.getFileFromState)
        } catch(error) {
          alert(error)
        }
        this.disabled = false
      },
      bundleClickHandler() {
        let bundleInput = document.getElementById("bundleHash")
        bundleInput.select()
        document.execCommand('copy')
      },
      ...mapActions('upload', [
        'setFileToState',
        'uploadFileToTangle',
        'resetBundleHash',
        'setUploadActive'
      ]),
      ...mapActions('download', [
        'setDownloadActive'
      ])
    },
    computed: {
      ...mapGetters('upload', [
        'getFileFromState',
        'getBundleHashFromState'
      ])
    }, 
    beforeMount() {
      this.setUploadActive(true)
      this.setDownloadActive(false)

      if (this.getBundleHashFromState) {
        this.uploadDisabled = true
      }
    }
  }
</script>

<style>
  .file-input {
    width: 400px;
    padding: 9px 5px;
  }
  .file-input-grp {
    display: inline;
  }
  button {
    border: 1px solid white;
    background: skyblue;
    padding: 10px 24px; 
    cursor: pointer;
    margin: 0;
  }
  button:hover {
    border-color: skyblue;
  }
  .bundle-hash {
    width: 750px;
    padding: 9px 5px;
    background-color: white;
  }
  button:disabled, button[disabled] {
    background: lightblue;
  }
</style>
