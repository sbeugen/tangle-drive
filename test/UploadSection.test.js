import { shallow, createLocalVue } from 'vue-test-utils'
import Vuex from 'vuex'
import Vue from 'vue'

import UploadSection from '../src/components/main_page/upload_section/UploadSection'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('UploadSection.test.js', () => {
  let store
  let uploadGetters
  let uploadActions
  let downloadActions
  let cmp

  uploadGetters = {
    getFileFromState: jest.fn().mockImplementation(() => ''),
    getBundleHashFromState: jest.fn().mockImplementation(() => '')
  }
  uploadActions = {
    setFileToState: jest.fn(),
    uploadFileToTangle: jest.fn(),
    resetBundleHash: jest.fn(),
    setUploadActive: jest.fn()
  }
  downloadActions = {
    setDownloadActive: jest.fn()
  }

  beforeEach(() => {
    store = new Vuex.Store({
      strict: true,
      modules: {
        upload: {
          namespaced: true,
          getters: uploadGetters,
          actions: uploadActions
        },
        download: {
          namespaced: true,
          actions: downloadActions
        }
      }
    })

    cmp = shallow(UploadSection, { store, localVue })
  })

  describe('Directly after mount', () => {
    it('has expected html structure', () => {
      expect(cmp.element).toMatchSnapshot()
    })

    it('Default values are set correctly', () => {
      expect(cmp.vm.disabled).toEqual(false)
      expect(cmp.vm.uploadDisabled).toEqual(false)
      expect(cmp.vm.MAX_FILE_SIZE).toBeGreaterThan(0)
    })

    it('functions inside beforeMount are called', () => {
      expect(uploadActions.setUploadActive).toBeCalled()
      expect(downloadActions.setDownloadActive).toBeCalled()
    })
  })

  describe('On clicking on File Button', () => {
    it('openFilePicker is called', () => {
      cmp.vm.openFilePicker = jest.fn()
      let fileButton = cmp.findAll('button').at(0)
      fileButton.trigger('click')
      expect(cmp.vm.openFilePicker).toBeCalled()
    })
    // Testing if fileInput is clicked is not working
    // it('openFilePicker clicks on fileInput -> ', () => {
    //   cmp.vm.fileInputClickHandler = jest.fn()
    //   let fileButton = cmp.findAll('button').at(0)
    //   fileButton.trigger('click')
    //   expect(cmp.vm.fileInputClickHandler).toBeCalled()
    // })
  })

  describe('On change of the fileInput', () => {

  })
})