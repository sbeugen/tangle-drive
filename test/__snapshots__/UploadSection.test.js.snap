// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadSection.test.js Directly after mount has expected html structure 1`] = `
<div>
  <p>
    1. Select a file you want to upload to the Tangle.
  </p>
   
  <div
    class="file-input-grp"
  >
    <input
      class="file-input"
      disabled="disabled"
      placeholder="Select a File"
      type="input"
    />
     
    <input
      hidden="hidden"
      id="file-selector"
      type="file"
    />
     
    <button>
      File
    </button>
     
    <br />
     
    <br />
     
    <!---->
     
    <!---->
  </div>
</div>
`;
